primitives/src/opcodes.rs

```diff
@@ -432,8 +435,338 @@ impl Opcode {
             _ => None,
         }
     }
+
+    pub fn new(code: u8) -> Self {
+        Self { code }
+    }
+
+    /// Returns the string representation of the opcode.
+    ///
+    /// This function maps the `Opcode`'s `code` value (a `u8`) to its corresponding  
+    /// Bitcoin Script opcode name.  
+    ///
+    /// ### Special Cases:
+    /// - `0x00` maps to `"OP_0"` (also known as `OP_FALSE`).
+    /// - `0x01..=0x4b` represents push operations and maps to `"OP_PUSHBYTES_N"`,  
+    ///   where `N` is the number of bytes pushed onto the stack.
+    /// - `0x4f` maps to `"OP_PUSHNUM_NEG1"` (pushing `-1` onto the stack).
+    /// - `0x51..=0x60` map to `"OP_PUSHNUM_1"` through `"OP_PUSHNUM_16"`,  
+    ///   corresponding to pushing numbers `1` through `16`.
+    /// - Any unknown or undefined opcode returns `"UNKNOWN_OPCODE"`.
+    ///
+    /// # Example
+    /// ```
+    /// use crate::bitcoin_primitives::opcodes::Opcode;
+    /// 
+    /// let op = Opcode::new(0x51);
+    /// assert_eq!(op.as_str(), "OP_PUSHNUM_1");
+    /// ```
+    pub fn as_str(&self) -> &'static str {
+        #[cfg(feature = "std")]
+        {
+            if let Some(name) = opcode_lookup().get(&self.code) {
+                return name;
+            }
+        }
+
+        #[cfg(not(feature = "std"))]
+        // return lookup_opcode(self.code);
+        {
+        if let Some(name) = OPCODE_LOOKUP.iter().find(|(c, _)| **c == self.code).map(|(_, s)| *s) {
```

Comment:
No iteration should be needed. There are exactly 255 opcodes ordered from lowest to highest, so this only needs index access. Also no `HashMap` is required.